parser/parser.go

```diff
@@ -79,6 +85,23 @@ func (p *Parser) parseNode() (n ast.Node, eof bool, err error) {
 	p.scry(1)
 	tok := p.lookahead[0]
 
+	// http://es5.github.io/#A.4
+	if tok.Type == token.LBrace {
```

Comment:
I remember from the compiler construction book that the idea of always consuming tokens from the lexer can help writing code that does not ends up in a infinite loop, but if the parseNode function was able to cope with an already present lookahead instead of panicing a lot of duplication could be avoided without having to use the openbraces.

Nothing against it per se, but I already have a similar problem to parse variables initializations and will need a similar logic x_x. Besides the possibility of a loop, what would be the others downsides on the parseNode being able to accept a lookahead > 0 ? It is similar as logic that re-stacks the token. Even tough it can end up in loops (re-stacking too), it seems OK to:

* look ahead
* check that you can't handle it
* step back and let someone else parse it

But I may be missing some fundamental concept =/

Reply:
Perhaps re-stacking tokens would be better because the code would be more uniform. But allowing it will introduce the possibility of loops on the parser, but we cope with possibility of loops all the time =P

Reply:
I dont see much problem. The enforcement of parseNode always resetting the lookahead was just an way to make sure the lookahead tokens were being corrected consumed. Because as `next` consumes from lexer, a prior (not consumed) lookahead could get lost. There maybe other ways to control this.

Feel free to change and we see how it goes.

Reply:
@tiago4orion after we merge what we have today we could try different designs to address the limitations =)